app/component/itinerary/navigator/hooks/useRealtimeLegs.js

```diff
@@ -194,29 +194,43 @@ const useRealtimeLegs = (relayEnvironment, initialLegs) => {
     );
 
     setTimeAndRealTimeLegs(prev => {
+      // Maps previous legs with fresh real time transit legs. If transit leg start is in the past according to previous state,
+      // the transit leg is marked as frozen to prevent the start from shifting in UI.
+      // rtLegMap does not contain legs that have ended in the past as they've been filtered before updates are queried
       const rtLegs = prev.realTimeLegs.map(l => {
+        const freezeStart = l.freezeStart || legTime(l.start) <= now;
         const rtLeg =
           l.legId && rtLegMap[l.legId] ? { ...rtLegMap[l.legId] } : null;
         if (rtLeg) {
+          // If start is frozen, the property is deleted to prevent it from affecting any views
+          if (freezeStart) {
+            delete rtLeg.start;
+          }
+
           return {
             ...l,
-            ...rtLeg,
+            ...rtLeg, // delete above prevent this from overwriting a previous, frozen state
             to: {
               ...l.to,
               vehicleRentalStation: rtLeg.to.vehicleRentalStation,
             },
+            freezeStart,
           };
         }
-
+        // Non-transit legs are kept unfrozen for now to allow them to be scaled or shifted
         return l;
       });
-      // shift non-transit-legs to match possibly changed transit legs
+
+      // Shift unfrozen, non-transit-legs to match possibly changed transit legs
       matchLegEnds(rtLegs, now);
+
+      // Freezes any leg.start|end in the past
       const rtLegsWithFreezes = rtLegs.map(l => ({
         ...l,
-        freezeStart: legTime(l.start) <= now,
-        freezeEnd: legTime(l.end) <= now,
+        freezeStart: l.freezeStart || legTime(l.start) <= now,
+        freezeEnd: l.freezeEnd || legTime(l.end) <= now,
       }));
+
```

Comment:
Two issues:
- Why constantly copy complete legs? Just assign the two freezing values to existing legs.
-  I wonder if it would be better to remove check for previous freeze state and  just  compare legTime with now.  Frozen times won't move and now does not go back, so simpler expression is enough. Anyway,  I understand that checking existing state is faster but somehow this  looks a bit confusing.